### PR TITLE
ambient: enable locality LB for waypoint

### PR DESCRIPTION
**Pull request by howardjohn**
Eventually we will enable this for arbitrary Services (https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/4444-service-routing-preference#summary), but for now just do waypoints and internalTrafficPolicy=local

### CODE REVIEW

**Review by jwendell — APPROVED** [review]

File: pilot/pkg/serviceregistry/kube/controller/ambient/services.go [hunk 1]
```diff
@@ -135,14 +136,38 @@ func (a *index) constructService(svc *v1.Service, w *Waypoint) *workloadapi.Serv
 		waypointAddress = a.getWaypointAddress(w)
 	}
 
+	var lb *workloadapi.LoadBalancing
+	if svc.Labels[constants.ManagedGatewayLabel] == constants.ManagedGatewayMeshControllerLabel {
```
Comment (stevenctl) [inline pilot/pkg/serviceregistry/kube/controller/ambient/services.go:140]: I don't think we can rely on this as our "is-waypoint" check for much longer.

**Review by stevenctl — APPROVED** [review]

File: pilot/pkg/serviceregistry/kube/controller/ambient/services.go [hunk 2]
```diff
@@ -135,14 +136,38 @@ func (a *index) constructService(svc *v1.Service, w *Waypoint) *workloadapi.Serv
...
+		// This is waypoint. Enable locality routing
+		lb = &workloadapi.LoadBalancing{
+			// Prefer endpoints in close zones, but allow spilling over to further endpoints where required.
+			RoutingPreference: []workloadapi.LoadBalancing_Scope{
+				workloadapi.LoadBalancing_NETWORK,
+				workloadapi.LoadBalancing_REGION,
+				workloadapi.LoadBalancing_ZONE,
+				workloadapi.LoadBalancing_SUBZONE,
+			},
+			Mode: workloadapi.LoadBalancing_FAILOVER,
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/services.go:150]: Is this good to always enable locality lb for the waypoint? I can think of cases where this is not the most optimized path

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go [hunk 3]
```diff
@@ -418,3 +426,41 @@ func constructServices(p *v1.Pod, services []model.ServiceInfo) map[string]*work
 	}
 	return res
 }
+
+func getPodLocality(ctx krt.HandlerContext, Nodes krt.Collection[*v1.Node], pod *v1.Pod) *workloadapi.Locality {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:430]: In previous processing, we respect `"istio-locality"` label first, then node labels

File: pilot/pkg/serviceregistry/kube/controller/ambient/services.go (hunk 2 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/services.go:150]: Currently we don't have an API that we want to expose for ambient so was wanting to default here. In the near future we can use the Kubernetes API

File: pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go (hunk 3 above)
Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambient/workloads.go:430]: Yeah I saw that as legacy and wasn't sure we actually want to carry it forward in ambient?

**Pull request merged**